## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [ ]:
from google.colab import files

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
df = pd.read_csv ("/content/drive/MyDrive/notebook7/life_expectancy_data.csv" )


FileNotFoundError: ignored

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [ ]:
expvida = pd.read_csv ( r'life_expectancy_data.csv')

In [ ]:
print(expvida)

          Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01    

3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [ ]:
# Obtengo el tipo de estructura de datos


In [ ]:
type(expvida)

pandas.core.frame.DataFrame

4) ¿Cuantas filas y columnas tiene `expvida`?

In [ ]:
# Uso .shape para obtener cantidad de filas y columnas


In [ ]:
expvida.shape

(2938, 22)

**Numero de filas**:

2938 filas

**Numero de columnas**:

22 columnas

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [ ]:
# Obtengo el nombre de las columnas


In [ ]:
expvida.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

6) Inspecciona las primeras 10 filas de `expvida`

In [ ]:
# Inspecciono las primeras filas


In [ ]:
expvida.head(10)

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   
5  Afghanistan  2010  Developing              58.8            279.0   
6  Afghanistan  2009  Developing              58.6            281.0   
7  Afghanistan  2008  Developing              58.1            287.0   
8  Afghanistan  2007  Developing              57.5            295.0   
9  Afghanistan  2006  Developing              57.3            295.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   
5             74     0.01               79.679367         66.0      1989  ...   
6             77     0.01               56.762217         63.0      2861  ...   
7             80     0.03               25.873925         64.0      1599  ...   
8             82     0.02               10.910156         63.0      1141  ...   
9             84     0.03               17.171518         64.0      1990  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   
5   66.0               9.20         66.0        0.1  553.328940   2883167.0   
6   63.0               9.42         63.0        0.1  445.893298    284331.0   
7   64.0               8.33         64.0        0.1  373.361116   2729431.0   
8   63.0               6.73         63.0        0.1  369.835796  26616792.0   
9   58.0               7.43         58.0        0.1  272.563770   2589345.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   
5                   18.4                 18.4   
6                   18.6                 18.7   
7                   18.8                 18.9   
8                   19.0                 19.1   
9                   19.2                 19.3   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  
5                            0.448        9.2  
6                            0.434        8.9  
7                            0.433        8.7  
8                            0.415        8.4  
9                            0.405        8.1  

[10 rows x 22 columns]

7) Y ahora las 10 ultimas.

In [ ]:
# Inspecciono las ultimas filas


In [ ]:
expvida.tail(10)

Country  Year      Status  Life expectancy   Adult Mortality  \
2928  Zimbabwe  2009  Developing              50.0            587.0   
2929  Zimbabwe  2008  Developing              48.2            632.0   
2930  Zimbabwe  2007  Developing              46.6             67.0   
2931  Zimbabwe  2006  Developing              45.4              7.0   
2932  Zimbabwe  2005  Developing              44.6            717.0   
2933  Zimbabwe  2004  Developing              44.3            723.0   
2934  Zimbabwe  2003  Developing              44.5            715.0   
2935  Zimbabwe  2002  Developing              44.8             73.0   
2936  Zimbabwe  2001  Developing              45.3            686.0   
2937  Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
2928             30     4.64                1.040021         73.0       853   
2929             30     3.56               20.843429         75.0         0   
2930             29     3.88               29.814566         72.0       242   
2931             28     4.57               34.262169         68.0       212   
2932             28     4.14                8.717409         65.0       420   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
2928  ...   69.0               6.26         73.0       18.1   65.824121   
2929  ...   75.0               4.96         75.0       20.5  325.678573   
2930  ...   73.0               4.47         73.0       23.7  396.998217   
2931  ...   71.0               5.12          7.0       26.8  414.796232   
2932  ...   69.0               6.44         68.0       30.3  444.765750   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
2928   1381599.0                    7.5                  7.4   
2929  13558469.0                    7.8                  7.8   
2930   1332999.0                    8.2                  8.2   
2931  13124267.0                    8.6                  8.6   
2932    129432.0                    9.0                  9.0   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
2928                            0.419        9.9  
2929                            0.421        9.7  
2930                            0.414        9.6  
2931                            0.408        9.5  
2932                            0.406        9.3  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937                            0.434        9.8  

[10 rows x 22 columns]

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [ ]:
# Tipo de datos en cada columna


In [ ]:
expvida.dtypes

Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object

9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [ ]:
# Obtengo la cantidad de datos faltantes en cada columna


In [ ]:
expvida.isna().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [ ]:
# Obtengo el maximo valor de datos faltantes y lo guardo en la variable masnan


In [ ]:
masnan = expvida.isna().sum().max()
print(masnan)

652


In [ ]:
# Obtengo la cantidad de filas total y divido el maximo valor de datos faltantes por este numero
# y lo multiplico por 100
filas = expvida.shape[0]
porcentaje_nan = (masnan / filas) *100

In [ ]:
# Imprimo el maximo porcentaje de datos faltantes. Uso f-strings.
print(f"porcentaje de datos faltantes: {round(porcentaje_nan,1)}%")


porcentaje de datos faltantes: 22.2%


- str.format()
- Template Class
- f-strings

#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [ ]:
# Renombro columnas del DataFrame
expvida.rename(columns={"Life expectancy " : "life_expectancy", "BMI " : "bmi","Meales " : "measles"}, inplace = True )


In [ ]:
# Chequeo que se hayan reemplazado correctamente
# no se reemplazaron
expvida.columns


Index(['Country', 'Year', 'Status', 'life_expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [ ]:
# Obtengo el promedio de la columna
expvida.describe ( life_expectancy )



NameError: ignored

In [ ]:
expvida ['life_expectancy'].mean()

69.22493169398908

13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [ ]:
# Obtengo el promedio por pais
prom_pais = expvida.groupby (['Country'])['life_expectancy'].mean()
print(prom_pais.to_string())


Country
Afghanistan                                             58.19375
Albania                                                 75.15625
Algeria                                                 73.61875
Angola                                                  49.01875
Antigua and Barbuda                                     75.05625
Argentina                                               75.15625
Armenia                                                 73.40000
Australia                                               81.81250
Austria                                                 81.48125
Azerbaijan                                              70.73125
Bahamas                                                 74.28750
Bahrain                                                 75.72500
Bangladesh                                              69.30000
Barbados                                                74.35625
Belarus                                                 69.90625
Belgium          

14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [ ]:
# Maxima expectativa de vida usando numpy
max_vida_numpy= np.max(expvida['life_expectancy'])


In [ ]:
import pandas as pd

In [ ]:
# Maxima expectativa de vida usando pandas

expvida.loc [(expvida['life_expectancy']== max_expvida_pandas )]



NameError: ignored

In [ ]:
filas_maximo_Pandas = expvida[expvida['life_expectancy'] == expectativa_alta_Pandas]


NameError: ignored

In [ ]:
# Minima expectativa de vida usando numpy


In [ ]:
# Minima expectativa de vida usando pandas


In [ ]:
# Muestro los resultados maximos


In [ ]:
# Muestro los resultados minimos


In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo


In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo


15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [ ]:
# Obtengo los valores unicos de la columna status


16) ¿Cuantos paises se ubican en cada una de esas categorias de "Status"? _**Pista**_: Lee acerca de la función [value_counts()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [ ]:
# Cuento cuantos valores hay de cada categoria
expvida['Status'].value_counts()


Developing    2426
Developed      512
Name: Status, dtype: int64

18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [ ]:
# Cuento cuantas filas tienen valores distintos de cero
expvida.loc[(expvida['Year']==2015) &(expvida['Measles ']>0)]


Country  Year      Status  life_expectancy  Adult Mortality  \
0     Afghanistan  2015  Developing             65.0            263.0   
32        Algeria  2015  Developing             75.6             19.0   
48         Angola  2015  Developing             52.4            335.0   
96        Armenia  2015  Developing             74.8            118.0   
112     Australia  2015   Developed             82.8             59.0   
...           ...   ...         ...              ...              ...   
2826   Uzbekistan  2015  Developing             69.4            184.0   
2842      Vanuatu  2015  Developing             72.0             13.0   
2874     Viet Nam  2015  Developing             76.0            127.0   
2890        Yemen  2015  Developing             65.7            224.0   
2906       Zambia  2015  Developing             61.8             33.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
32               21      NaN                0.000000         95.0        63   
48               66      NaN                0.000000         64.0       118   
96                1      NaN                0.000000         94.0        33   
112               1      NaN                0.000000         93.0        74   
...             ...      ...                     ...          ...       ...   
2826             15      NaN                0.000000         99.0        22   
2842              0      NaN                0.000000         64.0        39   
2874             28      NaN                0.000000         97.0       256   
2890             37      NaN                0.000000         69.0       468   
2906             27      NaN                0.000000          9.0         9   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS           GDP  \
0     ...    6.0               8.16         65.0        0.1    584.259210   
32    ...   95.0                NaN         95.0        0.1   4132.762920   
48    ...    7.0                NaN         64.0        1.9   3695.793748   
96    ...   96.0                NaN         94.0        0.1    369.654776   
112   ...   93.0                NaN         93.0        0.1  56554.387600   
...   ...    ...                ...          ...        ...           ...   
2826  ...   99.0                NaN         99.0        0.1   2137.576852   
2842  ...   65.0                NaN         64.0        0.1    285.834140   
2874  ...   97.0                NaN         97.0        0.1           NaN   
2890  ...   63.0                NaN         69.0        0.1           NaN   
2906  ...    9.0                NaN          9.0        4.1   1313.889646   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
32    39871528.0                    6.0                  5.8   
48     2785935.0                    8.3                  8.2   
96      291695.0                    2.1                  2.2   
112   23789338.0                    0.6                  0.6   
...          ...                    ...                  ...   
2826    312989.0                    3.0                  3.1   
2842     26463.0                    1.5                  1.4   
2874         NaN                   14.2                 14.5   
2890         NaN                   13.6                 13.4   
2906    161587.0                    6.3                  6.1   

      Income composition of resources  Schooling  
0                               0.479       10.1  
32                              0.743       14.4  
48                              0.531       11.4  
96                              0.741       12.7  
112                             0.937       20.4  
...                               ...        ...  
2826                            0.697       12.1  
2842                            0.598       10.8  
2874                           